Add support for sending telegram messages to topics

r-xyz opened the pull request:
Based on original PR #104059 by [jgresty](https://github.com/jgresty).
Did not manage to merge conflicts, so I remade the changes from scratch, including suggestions from previous PR reviews.

Topics were added to telegram groups in November 2022, and to the
telegram-bot library in version 20.0. They are a purely additive change
that is exposed by a single parameter `message_thread_id`. Not passing
this parameter will not change the behaviour from current.

This same parameter is used to send messages to threads and messages to
topics inside groups.

https://telegram.org/blog/topics-in-groups-collectible-usernames/it?setln=en#topics-in-groups

Fixes #81888
Fixes #91750

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #81888, #91750
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/29857

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

r-xyz committed:
Add support for sending telegram messages to topics
    Based on original PR #104059 by [jgresty](https://github.com/jgresty).
    Did not manage to merge conflicts, so I remade the changes from scratch, including suggestions from previous PR reviews.

    Topics were added to telegram groups in November 2022, and to the
    telegram-bot library in version 20.0. They are a purely additive change
    that is exposed by a single parameter `message_thread_id`. Not passing
    this parameter will not change the behaviour from current.

    This same parameter is used to send messages to threads and messages to
    topics inside groups.

    https://telegram.org/blog/topics-in-groups-collectible-usernames/it?setln=en#topics-in-groups

    Fixes #81888
    Fixes #91750

malkstar reviewed (approved):
Checked all the send methods updated have the kwarg and they do, looks good to me.

r-xyz commented:
I am not sure why the checks are stalled, CLA is signed but checks are still in pause.
@malkstar , I requested a re-review since before it was in draft state (hopefully re-approving while open will restart the checks). :shrug:

malkstar reviewed (approved):
LGTM

malkstar commented:
Think it needs a maintainer's approval to kick off the tests

timoei commented:
> Think it needs a maintainer's approval to kick off the tests

Hey @malkstar,
I'm looking forward for the new feature. Do you have any recommendations how to get the maintainers approval?

romprod commented:
> > Think it needs a maintainer's approval to kick off the tests
> 
> Hey @malkstar, I'm looking forward for the new feature. Do you have any recommendations how to get the maintainers approval?

Same.

I've split a group into different discussions so that I can receive alerts for different devices and set alerts, for example I'm wanting to set an Telegram notification for an 'Alarmo' discussion but I don't want an notification for zones being set etc.

malkstar commented:
> > Think it needs a maintainer's approval to kick off the tests
> 
> Hey @malkstar, I'm looking forward for the new feature. Do you have any recommendations how to get the maintainers approval?

No recommendations unfortunately, I've only had one PR merged, but there's a huge backlog of PRs to be reviewed and not so many reviewers.

mprobber commented:
Hi, thanks for drafting this PR - very much looking forward to this functionality.  Any update on this?

r-xyz commented:
Hi all,Thanks for showing interest.
Unluckily I have no idea on how to get maintainers attention. Still, if meanwhile anybody could double check the code, test it locally and review it, that could maybe help.

luca-angemi commented:
I've been running this code as a custom component for more than a month now. Works great. Hopefully it will get merged soon 🤞

luca-angemi commented:
Thumb up in the first post could help!

r-xyz commented:
@luca-angemi If you are running this same PR  as custom components (i.e. you forked my repo, not the one from parent PR), then I would suggest to please "Review changes" (green button) under  "[Files changed](https://github.com/home-assistant/core/pull/112715/files)" tab.
I am not sure, but maybe it helps maintainers to see if PR are actually validated by multiple users.

r-xyz commented:
@tofuSCHNITZEL sorry to ping you here. I do not want to push, on the contrary I can imagine the amount of work being carried over by Home Assistant developers (I can see the amount of issues and PR).
I just wanted to double check if there is anything missing to unlock this PR, since it has now been stalled for about one month.
Thanks :)

tofuSCHNITZEL reviewed (approved):
LGTM!

r-xyz commented:
@tofuSCHNITZEL Thanks for the review. 
Unfortunately the PR workflow is still awaiting approval by a maintainer.
<img src="https://github.com/home-assistant/core/assets/100710244/7cc7d9cf-5084-4843-bf9f-db799ec795d8" width="500">
The only requested changes were from @homeassistant bot, asking to sign the CLA, which I did. The PR has now tag `cla-signed` and I cannot sign it twice:

<img src="https://github.com/home-assistant/core/assets/100710244/0bb3a0d6-1fb9-41c7-876f-4a4f43f94a2a" width="500">

Not sure if I am missing something else. Please, would it be possible for a maintainer to manually unlock the workflow?

tofuSCHNITZEL commented:
I think you did all you could. maybe @MartinHjelmare could have a quick look?

r-xyz committed:
Merge branch 'dev' into telegram-threads

r-xyz commented:
Hi @MartinHjelmare,  would you have any suggestion on how to proceed? I already signed the CLA (which was the only required changes) but somehow the workflow is still stuck. The PR is still pending review by @homeassistant

timoei commented:
Does anyone still have hope that this PR will ever be merged?

r-xyz commented:
@bdraco Apologies for directly pinging you here, but this PR was stuck since 2 months.
It keeps showing as "Changes requested", though the only requested changes were to sign CLA (which I did immediately after the request). Somehow this is still not detected (and probably keeping the PR out of sight from the maintainers?).
Please let me know if there is anything else I can do to move it forward.
Thanks :)

bdraco committed:
Merge branch 'dev' into telegram-threads

docorreas reviewed (approved):
Looks good to me.

r-xyz commented:
thanks, @bdraco 
Unfortunately it seems it is still stuck awaiting review from @homeassistant 
Any hints on how to unlock?

emontnemery reviewed (approved):
LGTM, thanks @r-xyz 👍

emontnemery reviewed (changes requested):
The PR has merge conflicts and needs to be rebased.

Also, please update or add tests to `tests/components/telegram_bot/test_telegram_bot.py` which covers the new functionality

r-xyz committed:
Merge branch 'dev' into telegram-threads

r-xyz committed:
Merge branch 'home-assistant:dev' into telegram-threads

r-xyz committed:
telegram_bot: add tests for threads feature.

r-xyz commented:
Resolved conflicts and added additional test with threads.

OddanN commented:
@r-xyz Works locally, test is passed :)
 Thank you!

r-xyz committed:
Merge branch 'dev' into telegram-threads

r-xyz commented:
@bdraco @emontnemery
Requested changes addressed, but it still looks pending `change-requested`.
I also edited the description, but `has-tests` label is not added.

emontnemery commented on tests/components/telegram_bot/test_telegram_bot.py:
```diff
@@ -45,6 +46,27 @@ async def test_send_message(hass: HomeAssistant, webhook_platform) -> None:
+    )
+    await hass.async_block_till_done()
+
+    assert len(events) == 1
+    assert events[0].context == context
```
This doesn't show that the thread id is passed to the API.
The test already uses the fixture `mock_external_calls` indirectly via the `webhook_platform` fixture. I suggest to modify the `mock_external_calls` fixture to yield the patches instead of yielding nothing, then you can test here that `telegram.Bot.send_message` was called once with the correct data.

r-xyz replied:
Hi @emontnemery , Thanks for the heads up.
I added `message_thread_id` to event log (since it might actually be useful in automation or for debug) and added the corresponding test in tests.

luca-angemi commented:
hi @r-xyz there was a change in the init file for the latest beta. Please rebase when adding tests.

r-xyz committed:
Merge branch 'home-assistant:dev' into telegram-threads

r-xyz committed:
telegram_bot: fixed tests for threads.

r-xyz committed:
telegram_bot: fixed wrong line.

r-xyz commented:
I added `message_thread_id` to event log (since it might actually be useful in automation or for debug purposes) and added additional test [here](https://github.com/home-assistant/core/blob/2024c086c8b8f299745be9fe45d6365f752a0226/tests/components/telegram_bot/test_telegram_bot.py#L69).
This should address changes suggested by @emontnemery.
@homeassistant workflow still require manual unlock.

emontnemery reviewed (approved):
LGTM, thanks @r-xyz 👍

emontnemery committed:
Update test_telegram_bot.py
